### ISSUE TITLE
[test] Train-step fp16 tests flake — shared data/split fixture clobbered by real pipeline (vocab mismatch)

### ISSUE THREAD

**Issue by travisgalloway** [travisgalloway/monica#58]
## Symptom
`tests/test_mlx_train_step.py` fails **nondeterministically** — the fp16 tests
(`test_fp16_clean_step_updates_and_reports_scale`, `test_fp16_overflow_skips_update_and_backs_off`,
`test_grad_accum_two_identical_microbatches_equal_single`) flake, with *which* ones fail varying
run to run. The fp16 scaler reports a spurious `skipped=True` / non-finite `grad_norm` on the toy model.

## Root cause — shared data fixture, not flaky math
The test loads its batch from a **shared, gitignored path**:
```python
TRAIN_BIN = "data/split/train.bin"
def _first_batch(cfg):
    loader = PackedLoader(TRAIN_BIN, cfg.seq_len, batch_size=4, shuffle=False)
```
…using the **toy config (vocab_size = 256)**. But `data/split/` is the *same path the real
training pipeline writes to* (`src.data.split`). After a real POC data-prep + run, `data/split/train.bin`
holds **OLMo-tokenized uint16 ids (vocab 50280)**. Feeding ids up to 50279 into a vocab-256 model gives
out-of-bounds embedding/target lookups → garbage/non-finite logits → non-finite gradients → the fp16
scaler flags overflow and skips. MLX's out-of-bounds gather is undefined, so the read garbage (and thus
which assertions trip) varies between runs → looks like flakiness.

## Evidence
- A fresh in-range random batch (ids < vocab_size) makes the exact fp16 clean-step path **fully
  deterministic**: `skipped=False`, `grad_norm=1.139`, `loss=8.115`, identical across 8 trials.
- `data/split/train.meta.json` currently reads `{"dtype": "uint16", "n_tokens": 105000000}` (the POC
  training corpus) — confirming the fixture was clobbered by the real run.
- `mlx_train_step.py` and the test are byte-identical to the last green run, so this is not a code regression.

This is a **test-isolation/hygiene bug**: a unit test depends on a shared, mutable, gitignored data path
that the production pipeline overwrites. It will recur for anyone who runs the data pipeline, and any
other test reading `data/split/train.bin`/`data/smoke` is similarly fragile.

## Proposed fix
Make the train-step tests **hermetic** — don't read `data/split/`:
- Generate a tiny **byte-fallback / toy** dataset (vocab < 256) into a dedicated path or a pytest
  `tmp_path` fixture, OR build a small deterministic in-memory `(inputs, targets)` batch with
  `ids < cfg.vocab_size` (the 8-trial repro above is essentially this and is fully stable).
- Audit other tests for the same shared-path assumption; route any real-data smoke through an
  explicitly toy-scoped fixture.
- Optional guard: `PackedLoader`/the test could assert `max(id) < vocab_size` and fail loudly with a
  clear message instead of silently producing NaNs.

## Notes
- Not a perf item (distinct from #30/#32/#33/#34, which optimize the train step's speed) — same
  subsystem, so cross-listed here for visibility.
- Unrelated to PR #56 (generation/eval); surfaced while running the full suite during that PR's review.
